// @nrwl/devkit/src/utils/convert-nx-executor.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.convertNxExecutor = void 0;
const rxjs_1 = require("rxjs");
const workspace_1 = require("@nrwl/tao/src/shared/workspace");
/**
 * Convert an Nx Executor into an Angular Devkit Builder
 *
 * Use this to expose a compatible Angular Builder
 */
// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
function convertNxExecutor(executor) {
    const builderFunction = (options, builderContext) => {
        const workspaceConfig = new workspace_1.Workspaces(builderContext.workspaceRoot).readWorkspaceConfiguration();
        const context = {
            root: builderContext.workspaceRoot,
            projectName: builderContext.target.project,
            targetName: builderContext.target.target,
            configurationName: builderContext.target.configuration,
            workspace: workspaceConfig,
            cwd: process.cwd(),
            isVerbose: false,
        };
        if (builderContext.target &&
            builderContext.target.project &&
            builderContext.target.target) {
            context.target =
                workspaceConfig.projects[builderContext.target.project].targets[builderContext.target.target];
        }
        return toObservable(executor(options, context));
    };
    return require('@angular-devkit/architect').createBuilder(builderFunction);
}
exports.convertNxExecutor = convertNxExecutor;
function toObservable(promiseOrAsyncIterator) {
    if (typeof promiseOrAsyncIterator.then === 'function') {
        return rxjs_1.from(promiseOrAsyncIterator);
    }
    else {
        return new rxjs_1.Observable((subscriber) => {
            let asyncIterator = promiseOrAsyncIterator;
            function recurse(iterator) {
                iterator
                    .next()
                    .then((result) => {
                    if (!result.done) {
                        subscriber.next(result.value);
                        recurse(iterator);
                    }
                    else {
                        if (result.value) {
                            subscriber.next(result.value);
                        }
                        subscriber.complete();
                    }
                })
                    .catch((e) => {
                    subscriber.error(e);
                });
            }
            recurse(asyncIterator);
            return () => {
                asyncIterator.return();
            };
        });
    }
}
