// @nrwl/devkit/src/utils/get-workspace-layout.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getWorkspacePath = exports.getWorkspaceLayout = void 0;
const json_1 = require("./json");
/**
 * Returns workspace defaults. It includes defaults folders for apps and libs,
 * and the default scope.
 *
 * Example:
 *
 * ```typescript
 * { appsDir: 'apps', libsDir: 'libs', npmScope: 'myorg' }
 * ```
 * @param host - file system tree
 */
function getWorkspaceLayout(host) {
    var _a, _b, _c, _d;
    const nxJson = json_1.readJson(host, 'nx.json');
    const rawWorkspace = json_1.readJson(host, getWorkspacePath(host));
    return {
        appsDir: (_b = (_a = nxJson.workspaceLayout) === null || _a === void 0 ? void 0 : _a.appsDir) !== null && _b !== void 0 ? _b : 'apps',
        libsDir: (_d = (_c = nxJson.workspaceLayout) === null || _c === void 0 ? void 0 : _c.libsDir) !== null && _d !== void 0 ? _d : 'libs',
        npmScope: nxJson.npmScope,
        standaloneAsDefault: Object.values(rawWorkspace.projects).reduce(
        // default for second, third... projects should be based on all projects being defined as a path
        // for configuration read from ng schematics, this is determined by configFilePath's presence
        (allStandalone, next) => allStandalone && (typeof next === 'string' || 'configFilePath' in next), 
        // default for first project should be false
        Object.values(rawWorkspace.projects).length > 0),
    };
}
exports.getWorkspaceLayout = getWorkspaceLayout;
function getWorkspacePath(host) {
    const possibleFiles = ['/angular.json', '/workspace.json'];
    return possibleFiles.filter((path) => host.exists(path))[0];
}
exports.getWorkspacePath = getWorkspacePath;
