// @nrwl/devkit/src/utils/invoke-nx-generator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.convertNxGenerator = void 0;
const tslib_1 = require("tslib");
const path_1 = require("path");
class RunCallbackTask {
    constructor(callback) {
        this.callback = callback;
    }
    toConfiguration() {
        return {
            name: 'RunCallback',
            options: {
                callback: this.callback,
            },
        };
    }
}
function createRunCallbackTask() {
    return {
        name: 'RunCallback',
        create: () => {
            return Promise.resolve(({ callback }) => tslib_1.__awaiter(this, void 0, void 0, function* () {
                yield callback();
            }));
        },
    };
}
/**
 * Convert an Nx Generator into an Angular Devkit Schematic
 */
function convertNxGenerator(generator) {
    // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
    return (options) => invokeNxGenerator(generator, options);
}
exports.convertNxGenerator = convertNxGenerator;
/**
 * Create a Rule to invoke an Nx Generator
 */
function invokeNxGenerator(generator, options) {
    return (tree, context) => tslib_1.__awaiter(this, void 0, void 0, function* () {
        if (context.engine.workflow) {
            const engineHost = context.engine.workflow.engineHost;
            engineHost.registerTaskExecutor(createRunCallbackTask());
        }
        const root = context.engine.workflow && context.engine.workflow.engineHost.paths
            ? context.engine.workflow.engineHost.paths[1]
            : tree.root.path;
        const adapterTree = new DevkitTreeFromAngularDevkitTree(tree, root);
        const result = yield generator(adapterTree, options);
        if (!result) {
            return adapterTree['tree'];
        }
        if (typeof result === 'function') {
            if (context.engine.workflow) {
                context.addTask(new RunCallbackTask(result));
            }
        }
    });
}
const actionToFileChangeMap = {
    c: 'CREATE',
    o: 'UPDATE',
    d: 'DELETE',
};
class DevkitTreeFromAngularDevkitTree {
    constructor(tree, _root) {
        this.tree = tree;
        this._root = _root;
    }
    get root() {
        return this._root;
    }
    children(dirPath) {
        const { subdirs, subfiles } = this.tree.getDir(dirPath);
        return [...subdirs, ...subfiles];
    }
    delete(filePath) {
        this.tree.delete(filePath);
    }
    exists(filePath) {
        if (this.isFile(filePath)) {
            return this.tree.exists(filePath);
        }
        else {
            return this.children(filePath).length > 0;
        }
    }
    isFile(filePath) {
        return this.tree.exists(filePath) && !!this.tree.read(filePath);
    }
    listChanges() {
        const fileChanges = [];
        for (const action of this.tree.actions) {
            if (action.kind === 'r') {
                fileChanges.push({
                    path: this.normalize(action.to),
                    type: 'CREATE',
                    content: this.read(action.to),
                });
                fileChanges.push({
                    path: this.normalize(action.path),
                    type: 'DELETE',
                    content: null,
                });
            }
            else if (action.kind === 'c' || action.kind === 'o') {
                fileChanges.push({
                    path: this.normalize(action.path),
                    type: actionToFileChangeMap[action.kind],
                    content: action.content,
                });
            }
            else {
                fileChanges.push({
                    path: this.normalize(action.path),
                    type: 'DELETE',
                    content: null,
                });
            }
        }
        return fileChanges;
    }
    normalize(path) {
        return path_1.relative(this.root, path_1.join(this.root, path));
    }
    read(filePath, encoding) {
        return encoding
            ? this.tree.read(filePath).toString(encoding)
            : this.tree.read(filePath);
    }
    rename(from, to) {
        this.tree.rename(from, to);
    }
    write(filePath, content) {
        if (this.tree.exists(filePath)) {
            this.tree.overwrite(filePath, content);
        }
        else {
            this.tree.create(filePath, content);
        }
    }
}
