// carbon-components-angular/timepicker-select/timepicker-select.component.ngfactory.js
/*!
 *
 * carbon-angular v0.0.0 | timepicker-select.component.ngfactory.js
 *
 * Copyright 2014, 2019 IBM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0

 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */


/**
 * @fileoverview This file was generated by the Angular template compiler. Do not edit.
 *
 * @suppress {suspiciousCode,uselessCode,missingProperties,missingOverride,checkTypes}
 * tslint:disable
 */ 
import * as i0 from "@angular/core";
import * as i1 from "../../../node_modules/@carbon/icons-angular/lib/chevron--down/16.ngfactory";
import * as i2 from "@carbon/icons-angular/lib/chevron--down/16";
import * as i3 from "@angular/common";
import * as i4 from "./timepicker-select.component";
var styles_TimePickerSelect = [];
var RenderType_TimePickerSelect = i0.ɵcrt({ encapsulation: 2, styles: styles_TimePickerSelect, data: {} });
export { RenderType_TimePickerSelect as RenderType_TimePickerSelect };
function View_TimePickerSelect_1(_l) { return i0.ɵvid(0, [(_l()(), i0.ɵeld(0, 0, null, null, 1, "label", [["class", "bx--label bx--visually-hidden"]], [[1, "for", 0]], null, null, null, null)), (_l()(), i0.ɵted(1, null, ["", ""]))], null, function (_ck, _v) { var _co = _v.component; var currVal_0 = _co.id; _ck(_v, 0, 0, currVal_0); var currVal_1 = _co.label; _ck(_v, 1, 0, currVal_1); }); }
function View_TimePickerSelect_2(_l) { return i0.ɵvid(0, [(_l()(), i0.ɵeld(0, 0, null, null, 1, "ibm-icon-chevron-down16", [["class", "bx--select__arrow"]], null, null, null, i1.View_ChevronDown16_0, i1.RenderType_ChevronDown16)), i0.ɵdid(1, 49152, null, 0, i2.ChevronDown16, [], null, null)], null, null); }
export function View_TimePickerSelect_0(_l) { return i0.ɵvid(0, [i0.ɵqud(402653184, 1, { select: 0 }), (_l()(), i0.ɵand(16777216, null, null, 1, null, View_TimePickerSelect_1)), i0.ɵdid(2, 16384, null, 0, i3.NgIf, [i0.ViewContainerRef, i0.TemplateRef], { ngIf: [0, "ngIf"] }, null), (_l()(), i0.ɵeld(3, 0, [[1, 0], ["select", 1]], null, 1, "select", [["class", "bx--select-input"]], [[1, "id", 0], [8, "disabled", 0]], [[null, "change"]], function (_v, en, $event) { var ad = true; var _co = _v.component; if (("change" === en)) {
        var pd_0 = (_co.onChange($event) !== false);
        ad = (pd_0 && ad);
    } return ad; }, null, null)), i0.ɵncd(null, 0), (_l()(), i0.ɵand(16777216, null, null, 1, null, View_TimePickerSelect_2)), i0.ɵdid(6, 16384, null, 0, i3.NgIf, [i0.ViewContainerRef, i0.TemplateRef], { ngIf: [0, "ngIf"] }, null)], function (_ck, _v) { var _co = _v.component; var currVal_0 = (!_co.skeleton && _co.label); _ck(_v, 2, 0, currVal_0); var currVal_3 = !_co.skeleton; _ck(_v, 6, 0, currVal_3); }, function (_ck, _v) { var _co = _v.component; var currVal_1 = _co.id; var currVal_2 = _co.disabled; _ck(_v, 3, 0, currVal_1, currVal_2); }); }
export function View_TimePickerSelect_Host_0(_l) { return i0.ɵvid(0, [(_l()(), i0.ɵeld(0, 0, null, null, 1, "ibm-timepicker-select", [], [[2, "bx--select", null], [2, "bx--time-picker__select", null], [2, "bx--skeleton", null], [2, "bx--select--light", null]], [[null, "blur"]], function (_v, en, $event) { var ad = true; if (("blur" === en)) {
        var pd_0 = (i0.ɵnov(_v, 1).blur() !== false);
        ad = (pd_0 && ad);
    } return ad; }, View_TimePickerSelect_0, RenderType_TimePickerSelect)), i0.ɵdid(1, 49152, null, 0, i4.TimePickerSelect, [], null, null)], null, function (_ck, _v) { var currVal_0 = i0.ɵnov(_v, 1).timeSelect; var currVal_1 = i0.ɵnov(_v, 1).timePickerSelect; var currVal_2 = i0.ɵnov(_v, 1).timePickerSelectSkeleton; var currVal_3 = i0.ɵnov(_v, 1).timePickerSelectLight; _ck(_v, 0, 0, currVal_0, currVal_1, currVal_2, currVal_3); }); }
var TimePickerSelectNgFactory = i0.ɵccf("ibm-timepicker-select", i4.TimePickerSelect, View_TimePickerSelect_Host_0, { display: "display", label: "label", helperText: "helperText", invalidText: "invalidText", id: "id", disabled: "disabled", skeleton: "skeleton", invalid: "invalid", theme: "theme" }, { selected: "selected", valueChange: "valueChange" }, ["*"]);
export { TimePickerSelectNgFactory as TimePickerSelectNgFactory };
