// carbon-components-angular/esm2015/notification/toast.component.js
/**
 *
 * carbon-angular v0.0.0 | toast.component.js
 *
 * Copyright 2014, 2022 IBM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0

 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */


import { Component, Input, HostBinding } from "@angular/core";
import { Notification } from "./notification.component";
import { ExperimentalService } from "carbon-components-angular/experimental";
import { NotificationDisplayService } from "./notification-display.service";
import { I18n } from "carbon-components-angular/i18n";
/**
 * Toast messages are displayed toward the top of the UI and do not interrupt user’s work.
 *
 * [See demo](../../?path=/story/components-notification--toast)
 *
 * <example-url>../../iframe.html?id=components-notification--toast</example-url>
 */
export class Toast extends Notification {
    constructor(notificationDisplayService, i18n, experimental) {
        super(notificationDisplayService, i18n);
        this.notificationDisplayService = notificationDisplayService;
        this.i18n = i18n;
        this.experimental = experimental;
        this.toastID = `toast-${Toast.toastCount++}`;
        this.toastClass = true;
        this.role = "alert";
        // disable inline notification styles
        this.notificationClass = false;
    }
    get isError() { return this.notificationObj["type"] === "error"; }
    get isInfo() { return this.notificationObj["type"] === "info"; }
    get isSuccess() { return this.notificationObj["type"] === "success"; }
    get isWarning() { return this.notificationObj["type"] === "warning"; }
    get isLowContrast() { return this.notificationObj.lowContrast; }
    ngOnInit() {
        if (!this.notificationObj.closeLabel) {
            this.notificationObj.closeLabel = this.i18n.get().NOTIFICATION.CLOSE_BUTTON;
        }
    }
}
Toast.toastCount = 0;
Toast.decorators = [
    { type: Component, args: [{
                selector: "ibm-toast",
                template: `
		<svg
			ibmIcon="error--filled"
			size="20"
			*ngIf="notificationObj.type === 'error'"
			class="bx--toast-notification__icon">
		</svg>
		<svg
			ibmIcon="warning--filled"
			size="20"
			*ngIf="notificationObj.type === 'warning'"
			class="bx--toast-notification__icon">
		</svg>
		<svg
			ibmIcon="checkmark--filled"
			size="20"
			*ngIf="notificationObj.type === 'success'"
			class="bx--toast-notification__icon">
		</svg>
		<svg
			ibmIcon="information--filled"
			size="20"
			*ngIf="notificationObj.type === 'info'"
			class="bx--toast-notification__icon">
		</svg>
		<div class="bx--toast-notification__details">
			<h3 *ngIf="!notificationObj.template" ibmToastTitle [innerHTML]="notificationObj.title"></h3>
			<div *ngIf="!notificationObj.template" ibmToastSubtitle>
				<span [innerHTML]="notificationObj.subtitle"></span>
				<ng-container *ngFor="let link of notificationObj.links">
					<a ibmLink [href]="link.href"> {{link.text}}</a>
				</ng-container>
			</div>
			<p *ngIf="!notificationObj.template" ibmToastCaption [innerHTML]="notificationObj.caption"></p>
			<ng-container *ngTemplateOutlet="notificationObj.template; context: { $implicit: notificationObj}"></ng-container>
		</div>
		<button
			*ngIf="!isCloseHidden"
			class="bx--toast-notification__close-button"
			type="button"
			[attr.aria-label]="notificationObj.closeLabel"
			(click)="onClose()">
			<svg ibmIcon="close" size="16" class="bx--toast-notification__close-icon"></svg>
		</button>
	`
            }] }
];
/** @nocollapse */
Toast.ctorParameters = () => [
    { type: NotificationDisplayService },
    { type: I18n },
    { type: ExperimentalService }
];
Toast.propDecorators = {
    notificationObj: [{ type: Input }],
    toastID: [{ type: HostBinding, args: ["attr.id",] }],
    toastClass: [{ type: HostBinding, args: ["class.bx--toast-notification",] }],
    role: [{ type: HostBinding, args: ["attr.role",] }],
    isError: [{ type: HostBinding, args: ["class.bx--toast-notification--error",] }],
    isInfo: [{ type: HostBinding, args: ["class.bx--toast-notification--info",] }],
    isSuccess: [{ type: HostBinding, args: ["class.bx--toast-notification--success",] }],
    isWarning: [{ type: HostBinding, args: ["class.bx--toast-notification--warning",] }],
    isLowContrast: [{ type: HostBinding, args: ["class.bx--toast-notification--low-contrast",] }]
};
