// carbon-components-angular/esm5/pagination/pagination-nav/pagination-overflow.component.js
/**
 *
 * carbon-angular v0.0.0 | pagination-overflow.component.js
 *
 * Copyright 2014, 2022 IBM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0

 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */


import * as tslib_1 from "tslib";
import { Component, Input, Output, EventEmitter } from "@angular/core";
import { I18n } from "carbon-components-angular/i18n";
/**
 * Used to present a selection of pages when there is an overflow
 * in the pagination list
 *
 * * ```html
 * <ibm-pagination-overflow [fromIndex]="5" [count]="30" (change)="handleChange(value)"></ibm-pagination-overflow>
 * ```
 */
var PaginationOverflow = /** @class */ (function () {
    function PaginationOverflow(i18n) {
        this.i18n = i18n;
        this.ariaLabel = this.i18n.get().PAGINATION.SELECT_ARIA;
        /**
         * Emits click event
         */
        this.change = new EventEmitter();
    }
    Object.defineProperty(PaginationOverflow.prototype, "countAsArray", {
        get: function () {
            return tslib_1.__spread(Array(this.count));
        },
        enumerable: true,
        configurable: true
    });
    PaginationOverflow.prototype.handleChange = function (event) {
        this.change.emit(+event.target.value);
        event.target.value = "";
    };
    PaginationOverflow.decorators = [
        { type: Component, args: [{
                    selector: "ibm-pagination-overflow",
                    template: "\n\t\t<li class=\"bx--pagination-nav__list-item\" *ngIf=\"count > 1\">\n\t\t\t<div class=\"bx--pagination-nav__select\">\n\t\t\t<select\n\t\t\t\t[attr.aria-label]=\"ariaLabel\"\n\t\t\t\tclass=\"bx--pagination-nav__page bx--pagination-nav__page--select\"\n\t\t\t\t(change)=\"handleChange($event)\">\n\t\t\t\t<option value=\"\" hidden></option>\n\t\t\t\t<option\n\t\t\t\t*ngFor=\"let item of countAsArray; let i = index\">\n\t\t\t\t\t{{fromIndex + i + 1}}\n\t\t\t\t</option>\n\t\t\t</select>\n\t\t\t<div class=\"bx--pagination-nav__select-icon-wrapper\">\n\t\t\t\t<svg\n\t\t\t\t\tibmIcon=\"overflow-menu--horizontal\"\n\t\t\t\t\tsize=\"16\"\n\t\t\t\t\tstyle=\"display: inherit\"\n\t\t\t\t\tclass=\"bx--pagination-nav__select-icon\">\n\t\t\t\t</svg>\n\t\t\t</div>\n\t\t\t</div>\n\t\t</li>\n\t\t<ibm-pagination-nav-item *ngIf=\"count === 1\" [page]=\"fromIndex + 1\"></ibm-pagination-nav-item>\n\t"
                }] }
    ];
    /** @nocollapse */
    PaginationOverflow.ctorParameters = function () { return [
        { type: I18n }
    ]; };
    PaginationOverflow.propDecorators = {
        fromIndex: [{ type: Input }],
        count: [{ type: Input }],
        ariaLabel: [{ type: Input }],
        change: [{ type: Output }]
    };
    return PaginationOverflow;
}());
export { PaginationOverflow };
