// @nrwl/devkit/src/generators/project-configuration.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.readProjectConfiguration = exports.updateWorkspaceConfiguration = exports.readWorkspaceConfiguration = exports.getProjects = exports.removeProjectConfiguration = exports.updateProjectConfiguration = exports.addProjectConfiguration = void 0;
const tslib_1 = require("tslib");
const workspace_1 = require("@nrwl/tao/src/shared/workspace");
const json_1 = require("../utils/json");
const get_workspace_layout_1 = require("../utils/get-workspace-layout");
const path_1 = require("../utils/path");
/**
 * Adds project configuration to the Nx workspace.
 *
 * The project configuration is stored in workspace.json and nx.json. The utility will update
 * both files.
 *
 * @param host - the file system tree
 * @param projectName - unique name. Often directories are part of the name (e.g., mydir-mylib)
 * @param projectConfiguration - project configuration
 * @param standalone - should the project use package.json? If false, the project config is inside workspace.json
 */
function addProjectConfiguration(host, projectName, projectConfiguration, standalone = false) {
    standalone = standalone || get_workspace_layout_1.getWorkspaceLayout(host).standaloneAsDefault;
    setProjectConfiguration(host, projectName, projectConfiguration, 'create', standalone);
}
exports.addProjectConfiguration = addProjectConfiguration;
/**
 * Updates the configuration of an existing project.
 *
 * The project configuration is stored in workspace.json and nx.json. The utility will update
 * both files.
 *
 * @param host - the file system tree
 * @param projectName - unique name. Often directories are part of the name (e.g., mydir-mylib)
 * @param projectConfiguration - project configuration
 */
function updateProjectConfiguration(host, projectName, projectConfiguration) {
    setProjectConfiguration(host, projectName, projectConfiguration, 'update');
}
exports.updateProjectConfiguration = updateProjectConfiguration;
/**
 * Removes the configuration of an existing project.
 *
 * The project configuration is stored in workspace.json and nx.json.
 * The utility will update both files.
 */
function removeProjectConfiguration(host, projectName) {
    setProjectConfiguration(host, projectName, undefined, 'delete');
}
exports.removeProjectConfiguration = removeProjectConfiguration;
/**
 * Get a map of all projects in a workspace.
 *
 * Use {@link readProjectConfiguration} if only one project is needed.
 */
function getProjects(host) {
    const workspace = readWorkspace(host);
    const nxJson = json_1.readJson(host, 'nx.json');
    return new Map(Object.keys(workspace.projects || {}).map((projectName) => {
        return [
            projectName,
            getProjectConfiguration(host, projectName, workspace, nxJson),
        ];
    }));
}
exports.getProjects = getProjects;
/**
 * Read general workspace configuration such as the default project or cli settings
 *
 * This does _not_ provide projects configuration, use {@link readProjectConfiguration} instead.
 */
function readWorkspaceConfiguration(host) {
    const workspace = readWorkspace(host);
    delete workspace.projects;
    const nxJson = json_1.readJson(host, 'nx.json');
    delete nxJson.projects;
    return Object.assign(Object.assign({}, workspace), nxJson);
}
exports.readWorkspaceConfiguration = readWorkspaceConfiguration;
/**
 * Update general workspace configuration such as the default project or cli settings.
 *
 * This does _not_ update projects configuration, use {@link updateProjectConfiguration} or {@link addProjectConfiguration} instead.
 */
function updateWorkspaceConfiguration(host, workspaceConfig) {
    const { 
    // Angular Json Properties
    version, cli, defaultProject, generators, 
    // Nx Json Properties
    implicitDependencies, plugins, npmScope, targetDependencies, workspaceLayout, tasksRunnerOptions, affected, } = workspaceConfig;
    const workspace = {
        version,
        cli,
        defaultProject,
        generators,
    };
    const nxJson = {
        implicitDependencies,
        plugins,
        npmScope,
        targetDependencies,
        workspaceLayout,
        tasksRunnerOptions,
        affected,
    };
    json_1.updateJson(host, get_workspace_layout_1.getWorkspacePath(host), (json) => {
        return Object.assign(Object.assign({}, json), workspace);
    });
    json_1.updateJson(host, 'nx.json', (json) => {
        return Object.assign(Object.assign({}, json), nxJson);
    });
}
exports.updateWorkspaceConfiguration = updateWorkspaceConfiguration;
/**
 * Reads a project configuration.
 *
 * The project configuration is stored in workspace.json and nx.json. The utility will read
 * both files.
 *
 * @param host - the file system tree
 * @param projectName - unique name. Often directories are part of the name (e.g., mydir-mylib)
 * @throws If supplied projectName cannot be found
 */
function readProjectConfiguration(host, projectName) {
    const workspace = readWorkspace(host);
    if (!workspace.projects[projectName]) {
        throw new Error(`Cannot find configuration for '${projectName}' in ${get_workspace_layout_1.getWorkspacePath(host)}.`);
    }
    const nxJson = json_1.readJson(host, 'nx.json');
    // TODO: Remove after confirming that nx.json should be optional.
    // if (!nxJson.projects[projectName]) {
    //   throw new Error(
    //     `Cannot find configuration for '${projectName}' in nx.json`
    //   );
    // }
    return getProjectConfiguration(host, projectName, workspace, nxJson);
}
exports.readProjectConfiguration = readProjectConfiguration;
function getProjectConfiguration(host, projectName, workspace, nxJson) {
    return Object.assign(Object.assign({}, readWorkspaceSection(host, workspace, projectName)), readNxJsonSection(nxJson, projectName));
}
function readWorkspaceSection(host, workspace, projectName) {
    const config = workspace.projects[projectName];
    return config;
}
function readNxJsonSection(nxJson, projectName) {
    return nxJson.projects[projectName];
}
function setProjectConfiguration(host, projectName, projectConfiguration, mode, standalone = false) {
    if (mode === 'delete') {
        addProjectToNxJson(host, projectName, undefined, mode);
        addProjectToWorkspaceJson(host, projectName, undefined, mode);
        return;
    }
    if (!projectConfiguration) {
        throw new Error(`Cannot ${mode} "${projectName}" with value ${projectConfiguration}`);
    }
    const { tags, implicitDependencies } = projectConfiguration;
    addProjectToWorkspaceJson(host, projectName, projectConfiguration, mode, standalone);
    addProjectToNxJson(host, projectName, {
        tags,
        implicitDependencies,
    }, mode);
}
function addProjectToWorkspaceJson(host, projectName, project, mode, standalone = false) {
    const path = get_workspace_layout_1.getWorkspacePath(host);
    const workspaceJson = json_1.readJson(host, path);
    validateWorkspaceJsonOperations(mode, workspaceJson, projectName);
    const configFile = mode === 'create' && standalone
        ? path_1.joinPathFragments(project.root, 'project.json')
        : getProjectFileLocation(host, projectName);
    if (configFile) {
        if (mode === 'delete') {
            host.delete(configFile);
            delete workspaceJson.projects[projectName];
        }
        else {
            json_1.writeJson(host, configFile, project);
        }
        if (mode === 'create') {
            workspaceJson.projects[projectName] = project.root;
        }
    }
    else {
        let workspaceConfiguration;
        if (project) {
            const { tags, implicitDependencies } = project, c = tslib_1.__rest(project, ["tags", "implicitDependencies"]);
            workspaceConfiguration = c;
        }
        workspaceJson.projects[projectName] = workspaceConfiguration;
    }
    json_1.writeJson(host, path, workspaceJson);
}
function addProjectToNxJson(host, projectName, config, mode) {
    // distributed project files do not use nx.json,
    // so only proceed if the project does not use them.
    if (!getProjectFileLocation(host, projectName)) {
        const nxJson = json_1.readJson(host, 'nx.json');
        if (mode === 'delete') {
            delete nxJson.projects[projectName];
        }
        else {
            nxJson.projects[projectName] = Object.assign({
                tags: [],
            }, (config || {}));
        }
        json_1.writeJson(host, 'nx.json', nxJson);
    }
}
function readWorkspace(host) {
    const workspaceJson = inlineProjectConfigurationsWithTree(host);
    const originalVersion = workspaceJson.version;
    return Object.assign(Object.assign({}, workspace_1.toNewFormat(workspaceJson)), { version: originalVersion });
}
/**
 * This has to be separate from the inline functionality inside tao,
 * as the functionality in tao does not use a Tree. Changes made during
 * a generator would not be present during runtime execution.
 * @returns
 */
function inlineProjectConfigurationsWithTree(host) {
    const path = get_workspace_layout_1.getWorkspacePath(host);
    const workspaceJson = json_1.readJson(host, path);
    Object.entries(workspaceJson.projects || {}).forEach(([project, config]) => {
        if (typeof config === 'string') {
            const configFileLocation = path_1.joinPathFragments(config, 'project.json');
            workspaceJson.projects[project] = json_1.readJson(host, configFileLocation);
        }
    });
    return workspaceJson;
}
/**
 * @description Determine where a project's configuration is located.
 * @returns file path if separate from root config, null otherwise.
 */
function getProjectFileLocation(host, project) {
    var _a;
    const rawWorkspace = json_1.readJson(host, get_workspace_layout_1.getWorkspacePath(host));
    const projectConfig = (_a = rawWorkspace.projects) === null || _a === void 0 ? void 0 : _a[project];
    return typeof projectConfig === 'string'
        ? path_1.joinPathFragments(projectConfig, 'project.json')
        : null;
}
function validateWorkspaceJsonOperations(mode, workspaceJson, projectName) {
    if (mode == 'create' && workspaceJson.projects[projectName]) {
        throw new Error(`Cannot create Project '${projectName}'. It already exists.`);
    }
    if (mode == 'update' && !workspaceJson.projects[projectName]) {
        throw new Error(`Cannot update Project '${projectName}'. It does not exist.`);
    }
    if (mode == 'delete' && !workspaceJson.projects[projectName]) {
        throw new Error(`Cannot delete Project '${projectName}'. It does not exist.`);
    }
}
