// carbon-components-angular/dialog/overflow-menu/overflow-menu-option.component.ngfactory.js
/*!
 *
 * carbon-angular v0.0.0 | overflow-menu-option.component.ngfactory.js
 *
 * Copyright 2014, 2019 IBM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0

 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */


/**
 * @fileoverview This file was generated by the Angular template compiler. Do not edit.
 *
 * @suppress {suspiciousCode,uselessCode,missingProperties,missingOverride,checkTypes}
 * tslint:disable
 */ 
import * as i0 from "@angular/core";
import * as i1 from "@angular/common";
import * as i2 from "./overflow-menu-option.component";
var styles_OverflowMenuOption = [];
var RenderType_OverflowMenuOption = i0.ɵcrt({ encapsulation: 2, styles: styles_OverflowMenuOption, data: {} });
export { RenderType_OverflowMenuOption as RenderType_OverflowMenuOption };
function View_OverflowMenuOption_2(_l) { return i0.ɵvid(0, [(_l()(), i0.ɵeld(0, 0, null, null, 0, null, null, null, null, null, null, null))], null, null); }
function View_OverflowMenuOption_1(_l) { return i0.ɵvid(0, [(_l()(), i0.ɵeld(0, 0, null, null, 2, "button", [["class", "bx--overflow-menu-options__btn"], ["role", "menuitem"]], [[8, "tabIndex", 0], [8, "disabled", 0], [1, "title", 0]], [[null, "focus"], [null, "blur"], [null, "click"]], function (_v, en, $event) { var ad = true; var _co = _v.component; if (("focus" === en)) {
        var pd_0 = (_co.onFocus() !== false);
        ad = (pd_0 && ad);
    } if (("blur" === en)) {
        var pd_1 = (_co.onBlur() !== false);
        ad = (pd_1 && ad);
    } if (("click" === en)) {
        var pd_2 = (_co.onClick() !== false);
        ad = (pd_2 && ad);
    } return ad; }, null, null)), (_l()(), i0.ɵand(16777216, null, null, 1, null, View_OverflowMenuOption_2)), i0.ɵdid(2, 540672, null, 0, i1.NgTemplateOutlet, [i0.ViewContainerRef], { ngTemplateOutlet: [0, "ngTemplateOutlet"] }, null)], function (_ck, _v) { var currVal_3 = i0.ɵnov(_v.parent, 4); _ck(_v, 2, 0, currVal_3); }, function (_ck, _v) { var _co = _v.component; var currVal_0 = _co.tabIndex; var currVal_1 = _co.disabled; var currVal_2 = _co.title; _ck(_v, 0, 0, currVal_0, currVal_1, currVal_2); }); }
function View_OverflowMenuOption_4(_l) { return i0.ɵvid(0, [(_l()(), i0.ɵeld(0, 0, null, null, 0, null, null, null, null, null, null, null))], null, null); }
function View_OverflowMenuOption_3(_l) { return i0.ɵvid(0, [(_l()(), i0.ɵeld(0, 0, null, null, 2, "a", [["class", "bx--overflow-menu-options__btn"], ["role", "menuitem"]], [[8, "tabIndex", 0], [1, "disabled", 0], [8, "href", 4], [1, "title", 0]], [[null, "focus"], [null, "blur"], [null, "click"]], function (_v, en, $event) { var ad = true; var _co = _v.component; if (("focus" === en)) {
        var pd_0 = (_co.onFocus() !== false);
        ad = (pd_0 && ad);
    } if (("blur" === en)) {
        var pd_1 = (_co.onBlur() !== false);
        ad = (pd_1 && ad);
    } if (("click" === en)) {
        var pd_2 = (_co.onClick() !== false);
        ad = (pd_2 && ad);
    } return ad; }, null, null)), (_l()(), i0.ɵand(16777216, null, null, 1, null, View_OverflowMenuOption_4)), i0.ɵdid(2, 540672, null, 0, i1.NgTemplateOutlet, [i0.ViewContainerRef], { ngTemplateOutlet: [0, "ngTemplateOutlet"] }, null)], function (_ck, _v) { var currVal_4 = i0.ɵnov(_v.parent, 4); _ck(_v, 2, 0, currVal_4); }, function (_ck, _v) { var _co = _v.component; var currVal_0 = _co.tabIndex; var currVal_1 = _co.disabled; var currVal_2 = _co.href; var currVal_3 = _co.title; _ck(_v, 0, 0, currVal_0, currVal_1, currVal_2, currVal_3); }); }
function View_OverflowMenuOption_5(_l) { return i0.ɵvid(0, [(_l()(), i0.ɵeld(0, 0, null, null, 1, "div", [["class", "bx--overflow-menu-options__option-content"]], null, null, null, null, null)), i0.ɵncd(null, 0)], null, null); }
export function View_OverflowMenuOption_0(_l) { return i0.ɵvid(0, [(_l()(), i0.ɵand(16777216, null, null, 1, null, View_OverflowMenuOption_1)), i0.ɵdid(1, 16384, null, 0, i1.NgIf, [i0.ViewContainerRef, i0.TemplateRef], { ngIf: [0, "ngIf"] }, null), (_l()(), i0.ɵand(16777216, null, null, 1, null, View_OverflowMenuOption_3)), i0.ɵdid(3, 16384, null, 0, i1.NgIf, [i0.ViewContainerRef, i0.TemplateRef], { ngIf: [0, "ngIf"] }, null), (_l()(), i0.ɵand(0, [["tempOutlet", 2]], null, 0, null, View_OverflowMenuOption_5))], function (_ck, _v) { var _co = _v.component; var currVal_0 = !_co.href; _ck(_v, 1, 0, currVal_0); var currVal_1 = _co.href; _ck(_v, 3, 0, currVal_1); }, null); }
export function View_OverflowMenuOption_Host_0(_l) { return i0.ɵvid(0, [(_l()(), i0.ɵeld(0, 0, null, null, 1, "ibm-overflow-menu-option", [], [[8, "className", 0], [1, "role", 0], [2, "bx--overflow-menu-options__option--danger", null], [2, "bx--overflow-menu-options__option--disabled", null]], null, null, View_OverflowMenuOption_0, RenderType_OverflowMenuOption)), i0.ɵdid(1, 4243456, null, 0, i2.OverflowMenuOption, [i0.ElementRef], null, null)], null, function (_ck, _v) { var currVal_0 = i0.ɵnov(_v, 1).optionClass; var currVal_1 = i0.ɵnov(_v, 1).role; var currVal_2 = i0.ɵnov(_v, 1).isDanger; var currVal_3 = i0.ɵnov(_v, 1).isDisabled; _ck(_v, 0, 0, currVal_0, currVal_1, currVal_2, currVal_3); }); }
var OverflowMenuOptionNgFactory = i0.ɵccf("ibm-overflow-menu-option", i2.OverflowMenuOption, View_OverflowMenuOption_Host_0, { type: "type", disabled: "disabled", href: "href" }, { selected: "selected" }, ["*"]);
export { OverflowMenuOptionNgFactory as OverflowMenuOptionNgFactory };
