// carbon-components-angular/esm5/ui-shell/header/header-navigation.component.js
/**
 *
 * carbon-angular v0.0.0 | header-navigation.component.js
 *
 * Copyright 2014, 2021 IBM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0

 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */


import { Component, HostBinding, Input } from "@angular/core";
/**
 * Container for header navigation items
 */
var HeaderNavigation = /** @class */ (function () {
    function HeaderNavigation() {
        this.height = 100;
    }
    HeaderNavigation.decorators = [
        { type: Component, args: [{
                    selector: "ibm-header-navigation",
                    template: "\n\t\t<nav class=\"bx--header__nav\" [attr.aria-label]=\"ariaLabel\">\n\t\t\t<ul class=\"bx--header__menu-bar\" role=\"menubar\">\n\t\t\t\t<ng-content></ng-content>\n\t\t\t\t<ng-container *ngFor=\"let navigationItem of navigationItems\">\n\t\t\t\t\t<ibm-header-item\n\t\t\t\t\t\t*ngIf=\"navigationItem.type === 'item'\"\n\t\t\t\t\t\t[href]=\"navigationItem.href\"\n\t\t\t\t\t\t[route]=\"navigationItem.route\"\n\t\t\t\t\t\t[routeExtras]=\"navigationItem.routeExtras\"\n\t\t\t\t\t\t[isCurrentPage]=\"!!navigationItem.isCurrentPage\">\n\t\t\t\t\t\t{{ navigationItem.content }}\n\t\t\t\t\t</ibm-header-item>\n\t\t\t\t\t<ibm-header-menu\n\t\t\t\t\t\t*ngIf=\"navigationItem.type === 'menu'\"\n\t\t\t\t\t\t[href]=\"navigationItem.href\"\n\t\t\t\t\t\t[title]=\"navigationItem.title\"\n\t\t\t\t\t\t[trigger]=\"navigationItem.trigger ? navigationItem.trigger : 'click'\"\n\t\t\t\t\t\t[headerItems]=\"navigationItem.menuItems\">\n\t\t\t\t\t</ibm-header-menu>\n\t\t\t\t</ng-container>\n\t\t\t</ul>\n\t\t</nav>\n\t"
                }] }
    ];
    HeaderNavigation.propDecorators = {
        height: [{ type: HostBinding, args: ["style.height.%",] }],
        ariaLabel: [{ type: Input }],
        navigationItems: [{ type: Input }]
    };
    return HeaderNavigation;
}());
export { HeaderNavigation };
