// carbon-components-angular/esm2015/table/table-container.component.js
/**
 *
 * carbon-angular v0.0.0 | table-container.component.js
 *
 * Copyright 2014, 2022 IBM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0

 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */


import { Component, ContentChild, HostBinding } from "@angular/core";
import { TableHeaderDescription } from "./header/table-header-description.directive";
import { TableHeaderTitle } from "./header/table-header-title.directive";
import { Table } from "./table.component";
export class TableContainer {
    constructor() {
        this.containerClass = true;
    }
    ngAfterContentInit() {
        // Set aria properties if values exist otherwise keep undefined
        if (this.table) {
            // Using ternary instead of optional chaning since typescript version does not support it
            this.table.ariaLabelledby = this.headerTitle ? this.headerTitle.id : undefined;
            this.table.ariaDescribedby = this.headerDescription ? this.headerDescription.id : undefined;
        }
    }
}
TableContainer.decorators = [
    { type: Component, args: [{
                selector: "ibm-table-container",
                template: `<ng-content></ng-content>`,
                styles: [`
		:host { display: block }
	`]
            }] }
];
TableContainer.propDecorators = {
    containerClass: [{ type: HostBinding, args: ["class.bx--data-table-container",] }],
    headerTitle: [{ type: ContentChild, args: [TableHeaderTitle,] }],
    headerDescription: [{ type: ContentChild, args: [TableHeaderDescription,] }],
    table: [{ type: ContentChild, args: [Table,] }]
};
