// carbon-components-angular/table/cell/table-data.component.ngfactory.js
/*!
 *
 * carbon-angular v0.0.0 | table-data.component.ngfactory.js
 *
 * Copyright 2014, 2019 IBM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0

 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */


/**
 * @fileoverview This file was generated by the Angular template compiler. Do not edit.
 *
 * @suppress {suspiciousCode,uselessCode,missingProperties,missingOverride,checkTypes}
 * tslint:disable
 */ 
import * as i0 from "@angular/core";
import * as i1 from "@angular/common";
import * as i2 from "./table-data.component";
var styles_TableData = [];
var RenderType_TableData = i0.ɵcrt({ encapsulation: 2, styles: styles_TableData, data: {} });
export { RenderType_TableData as RenderType_TableData };
function View_TableData_1(_l) { return i0.ɵvid(0, [(_l()(), i0.ɵeld(0, 0, null, null, 1, null, null, null, null, null, null, null)), (_l()(), i0.ɵted(1, null, ["", ""]))], null, function (_ck, _v) { var _co = _v.component; var currVal_0 = _co.item.data; _ck(_v, 1, 0, currVal_0); }); }
function View_TableData_2(_l) { return i0.ɵvid(0, [(_l()(), i0.ɵand(0, null, null, 0))], null, null); }
export function View_TableData_0(_l) { return i0.ɵvid(0, [(_l()(), i0.ɵand(16777216, null, null, 1, null, View_TableData_1)), i0.ɵdid(1, 16384, null, 0, i1.NgIf, [i0.ViewContainerRef, i0.TemplateRef], { ngIf: [0, "ngIf"] }, null), (_l()(), i0.ɵand(16777216, null, null, 2, null, View_TableData_2)), i0.ɵdid(3, 540672, null, 0, i1.NgTemplateOutlet, [i0.ViewContainerRef], { ngTemplateOutletContext: [0, "ngTemplateOutletContext"], ngTemplateOutlet: [1, "ngTemplateOutlet"] }, null), i0.ɵpod(4, { data: 0 })], function (_ck, _v) { var _co = _v.component; var currVal_0 = !_co.item.template; _ck(_v, 1, 0, currVal_0); var currVal_1 = _ck(_v, 4, 0, _co.item.data); var currVal_2 = _co.item.template; _ck(_v, 3, 0, currVal_1, currVal_2); }, null); }
export function View_TableData_Host_0(_l) { return i0.ɵvid(0, [(_l()(), i0.ɵeld(0, 0, null, null, 1, "div", [["ibmTableData", ""]], null, null, null, View_TableData_0, RenderType_TableData)), i0.ɵdid(1, 49152, null, 0, i2.TableData, [], null, null)], null, null); }
var TableDataNgFactory = i0.ɵccf("[ibmTableData]", i2.TableData, View_TableData_Host_0, { item: "item", skeleton: "skeleton" }, {}, []);
export { TableDataNgFactory as TableDataNgFactory };
