// carbon-components-angular/esm2015/accordion/accordion.component.js
/**
 *
 * carbon-angular v0.0.0 | accordion.component.js
 *
 * Copyright 2014, 2022 IBM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0

 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */


import { Component, Input, ContentChildren, QueryList } from "@angular/core";
import { AccordionItem } from "./accordion-item.component";
/**
 * [See demo](../../?path=/story/components-accordion--basic)
 *
 * <example-url>../../iframe.html?id=components-accordion--basic</example-url>
 */
export class Accordion {
    constructor() {
        this.align = "end";
        this.size = "md";
        this._skeleton = false;
    }
    set skeleton(value) {
        this._skeleton = value;
        this.updateChildren();
    }
    get skeleton() {
        return this._skeleton;
    }
    ngAfterContentInit() {
        this.updateChildren();
    }
    updateChildren() {
        if (this.children) {
            this.children.toArray().forEach(child => child.skeleton = this.skeleton);
        }
    }
}
Accordion.decorators = [
    { type: Component, args: [{
                selector: "ibm-accordion",
                template: `
		<ul class="bx--accordion"
			[ngClass]="{
				'bx--accordion--end': align == 'end',
				'bx--accordion--start': align == 'start',
				'bx--accordion--sm': size == 'sm',
				'bx--accordion--md': size == 'md',
				'bx--accordion--lg': size == 'lg'
			}">
			<ng-content></ng-content>
		</ul>
	`
            }] }
];
Accordion.propDecorators = {
    align: [{ type: Input }],
    size: [{ type: Input }],
    children: [{ type: ContentChildren, args: [AccordionItem,] }],
    skeleton: [{ type: Input }]
};
