// carbon-components-angular/esm2015/forms/index.js
/**
 *
 * carbon-angular v0.0.0 | index.js
 *
 * Copyright 2014, 2021 IBM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0

 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */


// export the forms module
export { NFormsModule } from "./forms.module";
// re-export from form related modules and components
export { CheckboxModule, Checkbox, CheckboxChange, CheckboxState } from "carbon-components-angular/checkbox";
export { ToggleModule, Toggle, ToggleChange, ToggleState } from "carbon-components-angular/toggle";
export { RadioModule, Radio, RadioChange, RadioGroup } from "carbon-components-angular/radio";
export { InputModule, Label, TextArea, TextInput } from "carbon-components-angular/input";
export { ButtonModule, Button, ButtonSet } from "carbon-components-angular/button";
