// carbon-components-angular/breadcrumb/breadcrumb-item.component.ngfactory.js
/*!
 *
 * carbon-angular v0.0.0 | breadcrumb-item.component.ngfactory.js
 *
 * Copyright 2014, 2019 IBM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0

 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */


/**
 * @fileoverview This file was generated by the Angular template compiler. Do not edit.
 *
 * @suppress {suspiciousCode,uselessCode,missingProperties,missingOverride,checkTypes}
 * tslint:disable
 */ 
import * as i0 from "@angular/core";
import * as i1 from "@angular/common";
import * as i2 from "./breadcrumb-item.component";
var styles_BreadcrumbItemComponent = [];
var RenderType_BreadcrumbItemComponent = i0.ɵcrt({ encapsulation: 2, styles: styles_BreadcrumbItemComponent, data: {} });
export { RenderType_BreadcrumbItemComponent as RenderType_BreadcrumbItemComponent };
function View_BreadcrumbItemComponent_2(_l) { return i0.ɵvid(0, [(_l()(), i0.ɵeld(0, 0, null, null, 0, null, null, null, null, null, null, null))], null, null); }
function View_BreadcrumbItemComponent_1(_l) { return i0.ɵvid(0, [(_l()(), i0.ɵeld(0, 0, null, null, 2, "a", [["class", "bx--link"]], [[8, "href", 4], [1, "aria-current", 0]], null, null, null, null)), (_l()(), i0.ɵand(16777216, null, null, 1, null, View_BreadcrumbItemComponent_2)), i0.ɵdid(2, 540672, null, 0, i1.NgTemplateOutlet, [i0.ViewContainerRef], { ngTemplateOutlet: [0, "ngTemplateOutlet"] }, null)], function (_ck, _v) { var currVal_2 = i0.ɵnov(_v.parent, 2); _ck(_v, 2, 0, currVal_2); }, function (_ck, _v) { var _co = _v.component; var currVal_0 = i0.ɵinlineInterpolate(1, "", (_co.skeleton ? "/#" : _co.href), ""); var currVal_1 = (_co.current ? _co.ariaCurrent : null); _ck(_v, 0, 0, currVal_0, currVal_1); }); }
function View_BreadcrumbItemComponent_3(_l) { return i0.ɵvid(0, [i0.ɵncd(null, 0), (_l()(), i0.ɵand(0, null, null, 0))], null, null); }
export function View_BreadcrumbItemComponent_0(_l) { return i0.ɵvid(0, [(_l()(), i0.ɵand(16777216, null, null, 1, null, View_BreadcrumbItemComponent_1)), i0.ɵdid(1, 16384, null, 0, i1.NgIf, [i0.ViewContainerRef, i0.TemplateRef], { ngIf: [0, "ngIf"], ngIfElse: [1, "ngIfElse"] }, null), (_l()(), i0.ɵand(0, [["content", 2]], null, 0, null, View_BreadcrumbItemComponent_3))], function (_ck, _v) { var _co = _v.component; var currVal_0 = (_co.skeleton || _co.href); var currVal_1 = i0.ɵnov(_v, 2); _ck(_v, 1, 0, currVal_0, currVal_1); }, null); }
export function View_BreadcrumbItemComponent_Host_0(_l) { return i0.ɵvid(0, [(_l()(), i0.ɵeld(0, 0, null, null, 1, "ibm-breadcrumb-item", [], [[2, "bx--breadcrumb-item--current", null], [2, "bx--breadcrumb-item", null]], null, null, View_BreadcrumbItemComponent_0, RenderType_BreadcrumbItemComponent)), i0.ɵdid(1, 49152, null, 0, i2.BreadcrumbItemComponent, [], null, null)], null, function (_ck, _v) { var currVal_0 = i0.ɵnov(_v, 1).current; var currVal_1 = i0.ɵnov(_v, 1).itemClass; _ck(_v, 0, 0, currVal_0, currVal_1); }); }
var BreadcrumbItemComponentNgFactory = i0.ɵccf("ibm-breadcrumb-item", i2.BreadcrumbItemComponent, View_BreadcrumbItemComponent_Host_0, { href: "href", skeleton: "skeleton", ariaCurrent: "ariaCurrent", current: "current" }, {}, ["*"]);
export { BreadcrumbItemComponentNgFactory as BreadcrumbItemComponentNgFactory };
