// @angular-eslint/utils/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SelectorUtils = exports.Selectors = exports.RuleFixes = exports.ASTUtils = exports.getNativeEventNames = exports.getAriaAttributeKeys = exports.capitalize = exports.withoutBracketsAndWhitespaces = exports.kebabToCamelCase = exports.toPattern = exports.isNotNullOrUndefined = exports.arrayify = exports.toHumanReadableText = void 0;
__exportStar(require("./convert-annotated-source-to-failure-case"), exports);
__exportStar(require("./rules-tester"), exports);
var utils_1 = require("./utils");
Object.defineProperty(exports, "toHumanReadableText", { enumerable: true, get: function () { return utils_1.toHumanReadableText; } });
Object.defineProperty(exports, "arrayify", { enumerable: true, get: function () { return utils_1.arrayify; } });
Object.defineProperty(exports, "isNotNullOrUndefined", { enumerable: true, get: function () { return utils_1.isNotNullOrUndefined; } });
Object.defineProperty(exports, "toPattern", { enumerable: true, get: function () { return utils_1.toPattern; } });
Object.defineProperty(exports, "kebabToCamelCase", { enumerable: true, get: function () { return utils_1.kebabToCamelCase; } });
Object.defineProperty(exports, "withoutBracketsAndWhitespaces", { enumerable: true, get: function () { return utils_1.withoutBracketsAndWhitespaces; } });
Object.defineProperty(exports, "capitalize", { enumerable: true, get: function () { return utils_1.capitalize; } });
var get_aria_attribute_keys_1 = require("./eslint-plugin/get-aria-attribute-keys");
Object.defineProperty(exports, "getAriaAttributeKeys", { enumerable: true, get: function () { return get_aria_attribute_keys_1.getAriaAttributeKeys; } });
var get_native_event_names_1 = require("./eslint-plugin/get-native-event-names");
Object.defineProperty(exports, "getNativeEventNames", { enumerable: true, get: function () { return get_native_event_names_1.getNativeEventNames; } });
exports.ASTUtils = __importStar(require("./eslint-plugin/ast-utils"));
exports.RuleFixes = __importStar(require("./eslint-plugin/rule-fixes"));
exports.Selectors = __importStar(require("./eslint-plugin/selectors"));
exports.SelectorUtils = __importStar(require("./eslint-plugin/selector-utils"));
