// carbon-components-angular/esm5/tag/tag.component.js
/**
 *
 * carbon-angular v0.0.0 | tag.component.js
 *
 * Copyright 2014, 2022 IBM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0

 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */


import { Component, Input, HostBinding } from "@angular/core";
/**
 * Component that represents a tag for labelling/categorizing using keywords
 *
 * [See demo](../../?path=/story/components-tag--basic)
 *
 * <example-url>../../iframe.html?id=components-tag--basic</example-url>
 */
var Tag = /** @class */ (function () {
    function Tag() {
        /**
         * Type of the tag determines the styling
         */
        this.type = "gray";
        /**
         * Tag render size
         */
        this.size = "md";
        this.class = "";
    }
    Object.defineProperty(Tag.prototype, "attrClass", {
        get: function () {
            return "bx--tag bx--tag--" + this.type + " " + (this.size === "md" ? "bx--tag--md" : "bx--tag--sm") + " " + this.class;
        },
        enumerable: true,
        configurable: true
    });
    Tag.decorators = [
        { type: Component, args: [{
                    selector: "ibm-tag",
                    template: "<ng-content></ng-content>"
                }] }
    ];
    Tag.propDecorators = {
        type: [{ type: Input }],
        size: [{ type: Input }],
        class: [{ type: Input }],
        attrClass: [{ type: HostBinding, args: ["attr.class",] }]
    };
    return Tag;
}());
export { Tag };
