// carbon-components-angular/ui-shell/sidenav/sidenav-menu.component.js
/**
 *
 * carbon-angular v0.0.0 | sidenav-menu.component.js
 *
 * Copyright 2014, 2020 IBM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0

 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */


import { Component, ContentChildren, Input, QueryList } from "@angular/core";
import { SideNavItem } from "./sidenav-item.component";
import { Subscription } from 'rxjs';
/**
 * `SideNavMenu` provides a method to group `SideNavItem`s under a common heading.
 */
var SideNavMenu = /** @class */ (function () {
    function SideNavMenu() {
        /**
         * Controls the visibility of the child `SideNavItem`s
         */
        this.expanded = false;
        /**
         * Controls the active status indicator on the menu if there is an active
         * child sidenav item.
         */
        this.hasActiveChild = false;
        this.activeItemsSubscription = new Subscription();
    }
    SideNavMenu.prototype.findActiveChildren = function () {
        if (this.sidenavItems.some(function (item) { return item.active; })) {
            this.hasActiveChild = true;
        }
        else {
            this.hasActiveChild = false;
        }
    };
    SideNavMenu.prototype.ngAfterContentInit = function () {
        var _this = this;
        setTimeout(function () {
            _this.sidenavItems.forEach(function (item) {
                item.isSubMenu = true;
                _this.findActiveChildren();
                var activeItemSubscription = item.selected.subscribe(function () {
                    _this.findActiveChildren();
                });
                _this.activeItemsSubscription.add(activeItemSubscription);
            });
            _this.sidenavItems.changes.subscribe(function () {
                _this.sidenavItems.forEach(function (item) {
                    item.isSubMenu = true;
                    _this.findActiveChildren();
                    var activeItemSubscription = item.selected.subscribe(function () {
                        _this.findActiveChildren();
                    });
                    _this.activeItemsSubscription.add(activeItemSubscription);
                });
            });
        });
    };
    SideNavMenu.prototype.ngOnDestroy = function () {
        this.activeItemsSubscription.unsubscribe();
    };
    SideNavMenu.prototype.toggle = function () {
        this.expanded = !this.expanded;
    };
    SideNavMenu.decorators = [
        { type: Component, args: [{
                    selector: "ibm-sidenav-menu",
                    template: "\n\t\t<li\n\t\t\tclass=\"bx--side-nav__item bx--side-nav__item--icon\"\n\t\t\t[ngClass]=\"{ 'bx--side-nav__item--active': hasActiveChild }\">\n\t\t\t<button\n\t\t\t\t(click)=\"toggle()\"\n\t\t\t\tclass=\"bx--side-nav__submenu\"\n\t\t\t\taria-haspopup=\"true\"\n\t\t\t\t[attr.aria-expanded]=\"expanded\"\n\t\t\t\ttype=\"button\">\n\t\t\t\t<div class=\"bx--side-nav__icon\">\n\t\t\t\t\t<ng-content select=\"[icon]\"></ng-content>\n\t\t\t\t</div>\n\t\t\t\t<span class=\"bx--side-nav__submenu-title\">{{title}}</span>\n\t\t\t\t<div class=\"bx--side-nav__icon bx--side-nav__icon--small bx--side-nav__submenu-chevron\">\n\t\t\t\t\t<svg\n\t\t\t\t\t\tfocusable=\"false\"\n\t\t\t\t\t\tpreserveAspectRatio=\"xMidYMid meet\"\n\t\t\t\t\t\tstyle=\"will-change: transform;\"\n\t\t\t\t\t\txmlns=\"http://www.w3.org/2000/svg\"\n\t\t\t\t\t\twidth=\"20\"\n\t\t\t\t\t\theight=\"20\"\n\t\t\t\t\t\tviewBox=\"0 0 32 32\"\n\t\t\t\t\t\taria-hidden=\"true\">\n\t\t\t\t\t\t<path d=\"M16 22L6 12l1.4-1.4 8.6 8.6 8.6-8.6L26 12z\"></path>\n\t\t\t\t\t</svg>\n\t\t\t\t</div>\n\t\t\t</button>\n\t\t\t<ul class=\"bx--side-nav__menu\" role=\"menu\">\n\t\t\t\t<ng-content></ng-content>\n\t\t\t\t<ng-container *ngFor=\"let menuItem of menuItems\">\n\t\t\t\t\t<ibm-sidenav-item\n\t\t\t\t\t\t[href]=\"menuItem.href\"\n\t\t\t\t\t\t[route]=\"menuItem.route\"\n\t\t\t\t\t\t[routeExtras]=\"menuItem.routeExtras\"\n\t\t\t\t\t\t[isSubMenu]=\"true\">\n\t\t\t\t\t\t{{ menuItem.content }}\n\t\t\t\t\t</ibm-sidenav-item>\n\t\t\t\t</ng-container>\n\t\t\t</ul>\n\t\t</li>\n\t"
                },] },
    ];
    SideNavMenu.propDecorators = {
        title: [{ type: Input }],
        expanded: [{ type: Input }],
        hasActiveChild: [{ type: Input }],
        menuItems: [{ type: Input }],
        sidenavItems: [{ type: ContentChildren, args: [SideNavItem,] }]
    };
    return SideNavMenu;
}());
export { SideNavMenu };
