// carbon-components-angular/ui-shell/panel/panel.module.js
/*!
 *
 * carbon-angular v0.0.0 | panel.module.js
 *
 * Copyright 2014, 2019 IBM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0

 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */


import { NgModule } from "@angular/core";
import { CommonModule } from "@angular/common";
import { I18nModule } from "./../../i18n/i18n.module";
import { Panel } from "./panel.component";
import { ProductSwitcher } from "./product-switcher.component";
import { SwitcherList } from "./switcher-list.component";
import { SwitcherListItem } from "./switcher-list-item.component";
export { Panel, ProductSwitcher, SwitcherList, SwitcherListItem };
var PanelModule = /** @class */ (function () {
    function PanelModule() {
    }
    PanelModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [
                        Panel,
                        ProductSwitcher,
                        SwitcherList,
                        SwitcherListItem
                    ],
                    imports: [CommonModule, I18nModule],
                    exports: [
                        Panel,
                        ProductSwitcher,
                        SwitcherList,
                        SwitcherListItem
                    ]
                },] },
    ];
    return PanelModule;
}());
export { PanelModule };
