// carbon-components-angular/esm5/context-menu/context-menu-group.component.js
/**
 *
 * carbon-angular v0.0.0 | context-menu-group.component.js
 *
 * Copyright 2014, 2021 IBM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0

 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */


import { Component, EventEmitter, HostBinding, Input, Output } from "@angular/core";
import { Subscription } from "rxjs";
import { ContextMenuSelectionService } from "./context-menu-selection.service";
var ContextMenuGroupComponent = /** @class */ (function () {
    function ContextMenuGroupComponent(contextMenuSelectionService) {
        this.contextMenuSelectionService = contextMenuSelectionService;
        this.role = "none";
        this.label = null;
        this.value = [];
        this.type = null;
        this.valueChange = new EventEmitter();
        this.subscription = new Subscription();
    }
    ContextMenuGroupComponent.prototype.ngOnInit = function () {
        var _this = this;
        var selectionObservable = this.contextMenuSelectionService.selectionObservable;
        var subscription = selectionObservable.subscribe(function (value) {
            _this.valueChange.emit(value);
        });
        this.subscription.add(subscription);
    };
    ContextMenuGroupComponent.prototype.ngOnChanges = function (changes) {
        if (changes.value) {
            if (this.type === "radio") {
                this.contextMenuSelectionService.selectRadio(changes.value.currentValue);
            }
            if (this.type === "checkbox") {
                this.contextMenuSelectionService.selectCheckboxes(changes.value.currentValue);
            }
        }
    };
    ContextMenuGroupComponent.prototype.ngOnDestroy = function () {
        this.subscription.unsubscribe();
    };
    ContextMenuGroupComponent.decorators = [
        { type: Component, args: [{
                    selector: "ibm-context-menu-group",
                    template: "\n\t\t<ul role=\"group\" [attr.aria-label]=\"label\">\n\t\t\t<ng-content></ng-content>\n\t\t</ul>\n\t",
                    providers: [ContextMenuSelectionService],
                    styles: ["\n\t\t:host {\n\t\t\tdisplay: list-item;\n\t\t\tlist-style: none;\n\t\t}\n\t"]
                }] }
    ];
    /** @nocollapse */
    ContextMenuGroupComponent.ctorParameters = function () { return [
        { type: ContextMenuSelectionService }
    ]; };
    ContextMenuGroupComponent.propDecorators = {
        role: [{ type: HostBinding, args: ["attr.role",] }],
        label: [{ type: Input }],
        value: [{ type: Input }],
        type: [{ type: Input }],
        valueChange: [{ type: Output }]
    };
    return ContextMenuGroupComponent;
}());
export { ContextMenuGroupComponent };
