// carbon-components-angular/esm5/context-menu/context-menu-divider.component.js
/**
 *
 * carbon-angular v0.0.0 | context-menu-divider.component.js
 *
 * Copyright 2014, 2022 IBM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0

 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */


import { Component, HostBinding } from "@angular/core";
var ContextMenuDividerComponent = /** @class */ (function () {
    function ContextMenuDividerComponent() {
        this.dividerContextClass = true; // deprecated
        this.dividerClass = true;
        this.role = "separator";
    }
    ContextMenuDividerComponent.decorators = [
        { type: Component, args: [{
                    selector: "ibm-context-menu-divider",
                    template: "",
                    styles: ["\n\t\t:host {\n\t\t\tdisplay: list-item;\n\t\t\tlist-style: none;\n\t\t}\n\t"]
                }] }
    ];
    ContextMenuDividerComponent.propDecorators = {
        dividerContextClass: [{ type: HostBinding, args: ["class.bx--context-menu-divider",] }],
        dividerClass: [{ type: HostBinding, args: ["class.bx--menu-divider",] }],
        role: [{ type: HostBinding, args: ["attr.role",] }]
    };
    return ContextMenuDividerComponent;
}());
export { ContextMenuDividerComponent };
