// @angular/compiler-cli/src/ngtsc/typecheck/extended/api/api.js
/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define("@angular/compiler-cli/src/ngtsc/typecheck/extended/api/api", ["require", "exports", "tslib", "@angular/compiler", "@angular/compiler/src/compiler"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.TemplateCheckWithVisitor = void 0;
    var tslib_1 = require("tslib");
    var compiler_1 = require("@angular/compiler");
    var compiler_2 = require("@angular/compiler/src/compiler");
    /**
     * This abstract class provides a base implementation for the run method.
     */
    var TemplateCheckWithVisitor = /** @class */ (function () {
        function TemplateCheckWithVisitor() {
        }
        /**
         * Base implementation for run function, visits all nodes in template and calls
         * `visitNode()` for each one.
         */
        TemplateCheckWithVisitor.prototype.run = function (ctx, component, template) {
            var visitor = new TemplateVisitor(ctx, component, this);
            return visitor.getDiagnostics(template);
        };
        return TemplateCheckWithVisitor;
    }());
    exports.TemplateCheckWithVisitor = TemplateCheckWithVisitor;
    /**
     * Visits all nodes in a template (TmplAstNode and AST) and calls `visitNode` for each one.
     */
    var TemplateVisitor = /** @class */ (function (_super) {
        tslib_1.__extends(TemplateVisitor, _super);
        function TemplateVisitor(ctx, component, check) {
            var _this = _super.call(this) || this;
            _this.ctx = ctx;
            _this.component = component;
            _this.check = check;
            _this.diagnostics = [];
            return _this;
        }
        TemplateVisitor.prototype.visit = function (node, context) {
            var _a;
            (_a = this.diagnostics).push.apply(_a, tslib_1.__spreadArray([], tslib_1.__read(this.check.visitNode(this.ctx, this.component, node))));
            node.visit(this);
        };
        TemplateVisitor.prototype.visitAllNodes = function (nodes) {
            var e_1, _a;
            try {
                for (var nodes_1 = tslib_1.__values(nodes), nodes_1_1 = nodes_1.next(); !nodes_1_1.done; nodes_1_1 = nodes_1.next()) {
                    var node = nodes_1_1.value;
                    this.visit(node);
                }
            }
            catch (e_1_1) { e_1 = { error: e_1_1 }; }
            finally {
                try {
                    if (nodes_1_1 && !nodes_1_1.done && (_a = nodes_1.return)) _a.call(nodes_1);
                }
                finally { if (e_1) throw e_1.error; }
            }
        };
        TemplateVisitor.prototype.visitAst = function (ast) {
            if (ast instanceof compiler_2.ASTWithSource) {
                ast = ast.ast;
            }
            this.visit(ast);
        };
        TemplateVisitor.prototype.visitElement = function (element) {
            this.visitAllNodes(element.attributes);
            this.visitAllNodes(element.inputs);
            this.visitAllNodes(element.outputs);
            this.visitAllNodes(element.references);
            this.visitAllNodes(element.children);
        };
        TemplateVisitor.prototype.visitTemplate = function (template) {
            this.visitAllNodes(template.attributes);
            if (template.tagName === 'ng-template') {
                // Only visit input/outputs/templateAttrs if this isn't an inline template node
                // generated for a structural directive (like `<div *ngIf></div>`). These nodes
                // would be visited when the underlying element of an inline template node is processed.
                this.visitAllNodes(template.inputs);
                this.visitAllNodes(template.outputs);
                this.visitAllNodes(template.templateAttrs);
            }
            this.visitAllNodes(template.variables);
            this.visitAllNodes(template.references);
            this.visitAllNodes(template.children);
        };
        TemplateVisitor.prototype.visitContent = function (content) { };
        TemplateVisitor.prototype.visitVariable = function (variable) { };
        TemplateVisitor.prototype.visitReference = function (reference) { };
        TemplateVisitor.prototype.visitTextAttribute = function (attribute) { };
        TemplateVisitor.prototype.visitBoundAttribute = function (attribute) {
            this.visitAst(attribute.value);
        };
        TemplateVisitor.prototype.visitBoundEvent = function (attribute) {
            this.visitAst(attribute.handler);
        };
        TemplateVisitor.prototype.visitText = function (text) { };
        TemplateVisitor.prototype.visitBoundText = function (text) {
            this.visitAst(text.value);
        };
        TemplateVisitor.prototype.visitIcu = function (icu) { };
        TemplateVisitor.prototype.getDiagnostics = function (template) {
            this.diagnostics = [];
            this.visitAllNodes(template);
            return this.diagnostics;
        };
        return TemplateVisitor;
    }(compiler_1.RecursiveAstVisitor));
});
