// carbon-components-angular/esm5/context-menu/context-menu-selection.service.js
/**
 *
 * carbon-angular v0.0.0 | context-menu-selection.service.js
 *
 * Copyright 2014, 2021 IBM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0

 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */


import { Injectable } from "@angular/core";
import { ReplaySubject } from "rxjs";
var ContextMenuSelectionService = /** @class */ (function () {
    function ContextMenuSelectionService() {
        this.selectionSubject = new ReplaySubject(1);
        this.value = [];
        this.selectionObservable = this.selectionSubject.asObservable();
    }
    ContextMenuSelectionService.prototype.selectRadio = function (value) {
        if (!value) {
            return;
        }
        this.selectionSubject.next(value);
        this.value = [value];
    };
    ContextMenuSelectionService.prototype.selectCheckbox = function (value) {
        if (!value) {
            return;
        }
        if (this.value.includes(value)) {
            this.value = this.value.filter(function (v) { return v !== value; });
        }
        else {
            this.value.push(value);
        }
        this.selectionSubject.next(this.value);
    };
    ContextMenuSelectionService.prototype.selectCheckboxes = function (value) {
        if (!value) {
            return;
        }
        this.value = value;
        this.selectionSubject.next(value);
    };
    ContextMenuSelectionService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    ContextMenuSelectionService.ctorParameters = function () { return []; };
    return ContextMenuSelectionService;
}());
export { ContextMenuSelectionService };
